From 8d5cc67ac8ba4040cda2666ebcc903cf7abee927 Mon Sep 17 00:00:00 2001
From: Abhijeet Kaur <abkaur@google.com>
Date: Wed, 15 Jun 2022 17:01:29 +0100
Subject: [PATCH] isDataOrObbPath blocks access to Android/[data|obb] dirs only

isDataOrObbPath is a check used to restrict apps from accessing
Android/data and Android/obb directories via FUSE. However, if the app
is trying to access another directory in Android/data or Android/obb,
then we should not return early and should check if they have access to
the app package directory.

This only affects secondary volumes, as primary volumes Android/data and
Android/obb directories do not go through FUSE.

Bug: 201667614
Test: atest FileUtilsTest
Change-Id: I9d73c8c1ed40efb8bc851e9f510d4f7453367308
Merged-In: I9d73c8c1ed40efb8bc851e9f510d4f7453367308
(cherry picked from commit fe79a43a890d9c54655b0ad0beeab58958aa1cfb)
(cherry picked from commit bb4ec053ea16931e827f30c7b697cb35185a32be)
Merged-In: I9d73c8c1ed40efb8bc851e9f510d4f7453367308
---
 src/com/android/providers/media/util/FileUtils.java |  5 +++--
 .../android/providers/media/util/FileUtilsTest.java | 13 +++++++++----
 2 files changed, 12 insertions(+), 6 deletions(-)

diff --git a/src/com/android/providers/media/util/FileUtils.java b/src/com/android/providers/media/util/FileUtils.java
index 5c5e6b1c..a19d1a95 100644
--- a/src/com/android/providers/media/util/FileUtils.java
+++ b/src/com/android/providers/media/util/FileUtils.java
@@ -973,12 +973,13 @@ public class FileUtils {
             "(?i)^Android/(?:data|media|obb)/([^/]+)(/?.*)?");
 
     /**
-     * Regex that matches Android/obb or Android/data path.
+     * Regex that matches exactly Android/obb or Android/data or Android/obb/ or Android/data/
+     * suffix absolute file path.
      */
     private static final Pattern PATTERN_DATA_OR_OBB_PATH = Pattern.compile(
             "(?i)^/storage/[^/]+/(?:[0-9]+/)?"
             + PROP_CROSS_USER_ROOT_PATTERN
-            + "Android/(?:data|obb)(?:/.*)?$");
+            + "Android/(?:data|obb)/?$");
 
     /**
      * Regex that matches Android/obb or Android/data relative path (as defined in
diff --git a/tests/src/com/android/providers/media/util/FileUtilsTest.java b/tests/src/com/android/providers/media/util/FileUtilsTest.java
index 3501245d..ed0f5b6a 100644
--- a/tests/src/com/android/providers/media/util/FileUtilsTest.java
+++ b/tests/src/com/android/providers/media/util/FileUtilsTest.java
@@ -904,11 +904,16 @@ public class FileUtilsTest {
         assertThat(isDataOrObbPath("/storage/emulated/0/Android/obb")).isTrue();
         assertThat(isDataOrObbPath("/storage/ABCD-1234/Android/data")).isTrue();
         assertThat(isDataOrObbPath("/storage/ABCD-1234/Android/obb")).isTrue();
-        assertThat(isDataOrObbPath("/storage/emulated/0/Android/data/foo")).isTrue();
-        assertThat(isDataOrObbPath("/storage/emulated/0/Android/obb/foo")).isTrue();
-        assertThat(isDataOrObbPath("/storage/ABCD-1234/Android/data/foo")).isTrue();
-        assertThat(isDataOrObbPath("/storage/ABCD-1234/Android/obb/foo")).isTrue();
 
+        assertThat(isDataOrObbPath("/storage/emulated/0/Android/data/foo")).isFalse();
+        assertThat(isDataOrObbPath("/storage/emulated/0/Android/obb/foo")).isFalse();
+        assertThat(isDataOrObbPath("/storage/ABCD-1234/Android/data/foo")).isFalse();
+        assertThat(isDataOrObbPath("/storage/ABCD-1234/Android/obb/foo")).isFalse();
+        assertThat(isDataOrObbPath("/storage/emulated/10/Android/obb/foo")).isFalse();
+        assertThat(isDataOrObbPath("/storage/emulated//Android/obb/foo")).isFalse();
+        assertThat(isDataOrObbPath("/storage/emulated//Android/obb")).isFalse();
+        assertThat(isDataOrObbPath("/storage/emulated/0//Android/obb")).isFalse();
+        assertThat(isDataOrObbPath("/storage/emulated/0//Android/obb/foo")).isFalse();
         assertThat(isDataOrObbPath("/storage/emulated/0/Android/")).isFalse();
         assertThat(isDataOrObbPath("/storage/emulated/0/Android/media/")).isFalse();
         assertThat(isDataOrObbPath("/storage/ABCD-1234/Android/media/")).isFalse();
-- 
2.38.1.273.g43a17bfeac-goog

